File: CN/go_back_n/receiver.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <cstdlib>
#pragma comment(lib, "ws2_32.lib")

using namespace std;

int main() {
    WSADATA wsa;
    WSAStartup(MAKEWORD(2,2), &wsa);

    SOCKET sockfd;
    char buffer[1024];
    sockaddr_in receiverAddr{}, senderAddr{};
    int addr_len = sizeof(senderAddr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    receiverAddr.sin_family = AF_INET;
    receiverAddr.sin_port = htons(5001);
    receiverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    bind(sockfd, (sockaddr*)&receiverAddr, sizeof(receiverAddr));

    int expected = 0;

    cout << "RECEIVER STARTED (GO-BACK-N)\n\n";

    while (true) {
        memset(buffer, 0, sizeof(buffer));

        recvfrom(sockfd, buffer, sizeof(buffer), 0,
                 (sockaddr*)&senderAddr, &addr_len);

        cout << "RECEIVER: Frame arrived -> " << buffer << endl;

        int seq = stoi(string(buffer).substr(6, 1)); // FRAME:x

        // simulate frame loss
        if (rand() % 100 < 20) {
            cout << "RECEIVER: Frame lost\n\n";
            continue;
        }

        if (seq == expected) {
            cout << "RECEIVER: Accepted frame " << seq << endl;
            expected++;

            // cumulative ACK
            int ackNum = expected - 1;
            char ack[20];
            sprintf(ack, "ACK:%d", ackNum);

            // simulate ACK loss
            if (rand() % 100 < 20) {
                cout << "RECEIVER: ACK lost\n\n";
                continue;
            }

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: Sent cumulative ACK -> " << ack << endl;
        } else {
            cout << "RECEIVER: Out-of-order frame discarded\n";

            int ackNum = expected - 1;
            char ack[20];
            sprintf(ack, "ACK:%d", ackNum);

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: Re-sent ACK -> " << ack << endl;
        }

        cout << endl;
    }

    closesocket(sockfd);
    WSACleanup();
    return 0;
}

File: PRAC_EXAM/CN/4_DLL/go_back_n/receiver.py
import socket
import random

HOST = "127.0.0.1"
PORT = 5100

LOSS_RATE = 0.2
ACK_LOSS_RATE = 0.2

server = socket.socket()
server.bind((HOST, PORT))
server.listen(1)

print("RECEIVER STARTED\n")

conn, addr = server.accept()
print("Connected:", addr)

expected = 0

while True:

    data = conn.recv(1024)

    if not data:
        break

    frame = data.decode()
    print("Received:", frame)

    seq = int(frame.split(":")[1])

    # Simulate frame loss
    if random.random() < LOSS_RATE:
        print("Frame lost\n")
        continue

    if seq == expected:
        print("Accepted frame", seq)
        expected += 1
    else:
        print("Out-of-order frame discarded")

    ack = f"ACK:{expected-1}"

    # Simulate ACK loss
    if random.random() < ACK_LOSS_RATE:
        print("ACK lost\n")
        continue

    conn.send(ack.encode())
    print("Sent:", ack, "\n")

conn.close()
server.close()